### ISSUE TITLE
uints in CStructs (using NativeCall) aren't unsigned

### ISSUE THREAD

**Issue by jmaslak** [rakudo/rakudo#3851]
## The Problem

I'm not actually sure where to report NativeCall bugs.  My apology if this is the wrong place.

```
use v6.d;
use NativeCall;
class simple is repr('CStruct') {
    has uint16 $.value;
}

my $simple = simple.new(:value(0xffff));
say "BROKEN!" if $simple.value ≠ 0xffff;
```

## Expected Behavior

No output.  ;)  If getting the value of a uint16 (or other size of uint) in a CStruct, it's value should always be unsigned.

## Actual Behavior

In this case, $simple.value has a value of -1.

## Steps to Reproduce

See above.

## Environment
* Operating system: Linux
* Compiler version (`perl6 -v` or `raku -v`): 2020.05 & 2020.07 at least.

**Comment by JJ** [comment]
But do they behave natively as unsigned?

**Comment by jmaslak** [comment]
I'm not actually sure how to test how they behave natively.  I can write to them with large integers just fine (as the code demonstrates), but reading from them gives signed results.

**Comment by kjpye** [comment]
This is perhaps a MoarVM issue...

I have run into this issue using NativeCall to mmap a file -- I often need to randomly access large (hundreds of megabytes) files, so I mmap them and so can index directly where I need to go. Thus I have code like...

```
use NativeCall;

sub mmap(Pointer $addr, int32 $length, int32 $prot, int32 $flahs, int32 $fd, int32 $offset) returns CArray[uint8] is native { * }

my $fd = 'filename'.IO.open: :bin;
my $length = $filename.IO.s;
$file = mmap(Pointer, $length, 1, 1, $fd.native-descriptor, 0);
```
and I can then access random bytes in the file as `$file[$index]`.

Note that I have defined the mmapped file as `CArray[uint8]`, and so should get unsigned bytes -- but I don't, they are signed. This can be worked around, but is most inconvenient.

I have spent a couple of hours wandering further into the guts of MoarVM than I thought I ever would. I am not a MoarVM expert, so take the following with a grain of salt or two. I could be completely wrong.

The code to access the contents of CArrays is in src/6model/reprs/CArray.c, in particular in the routine `at_pos`. This code boils down to (in the case we're concerned with)

```
static void at_pos(MVMThreadContext *tc, MVMSTable *st, MVMObject *root, void *data, MVMint64 index, MVMRegister *value, MVMuint16 kind) {
    MVMCArrayREPRData *repr_data = (MVMCArrayREPRData *)st->REPR_data;
    MVMCArrayBody     *body      = (MVMCArrayBody *)data;
    void              *ptr       = ((char *)body->storage) + index * repr_data->elem_size;
    switch (repr_data->elem_kind) {
        case MVM_CARRAY_ELEM_KIND_NUMERIC:
            if (kind == MVM_reg_int64)
                value->i64 = body->managed && index >= body->elems
                    ? 0
                    : REPR(repr_data->elem_type)->box_funcs.get_int(tc,
                        STABLE(repr_data->elem_type), root, ptr);
            else ...
            break;
        ...
}

```
which just calls `REPR(repr_data->elem_type)->box_funcs.get_int` to get the value. There is an equivalent `get_uint` function available, but there is no code here to use it.

There is a `compose` routine in the same file which is presumably used to create the CArray structure in the first place. This creates just three pieces of information, all used in the code above -- the size of each element, the type of each element (numeric, string,...) and a pointer to the "type" which includes a pointer to the get_int routine. The information available to compose includes an "unsigned" flag, but this is ignored, even if it contains reliable information. We are reliant on the routine which calls at_pos (`MVM_repr_at_pos_i`) to specify we want an integer. `MVM_repr_at_pos_i` I assume is called from the relevant MoarVM opcode generated by the Rakudo compiler.

Similarly, in src/6model/reprs/CStruct.c there are references to `get_int`, but not to `get_uint`, so I assume the underlying problem is similar.

So to summarise -- there is no code in MoarVM to return unsigned ints from CArrays or CStructs.

**Comment by MasterDuke17** [comment]
I don't know NativeCall, you'll need @niner for that, but I do know Rakudo/MoarVM have a bunch of bugs/stuff-not-yet-implemented around unsigned native types. A couple years ago now I started a branch in MoarVM to add `*_u` versions of a lot of different functions/ops (https://github.com/MasterDuke17/MoarVM/tree/add_unsigned_int_ops_where_missing), but I don't remember if I ran into problems or just out of steam.

**Comment by niner** [comment]
I don't think that using get_int instead of get_uint makes any difference 
here. Both will copy the same number of bits unmodified. What makes an integer 
signed or unsigned is how those bits are interpreted. I suspect the visible 
errors are due to something going wrong at a higher level. You could confirm 
this by just replacing usage of get_int with get_uint in MoarVM and checking 
if that changes anything. I suspect it doesn't.

A quick glance at lib/NativeCall/Types.rakumod shows that IntTypedCArray uses 
nqp::atposref_i and nqp::bindpos_i. There's no distinction between signed and 
unsigned. Indeed, we do not have a nqp::atposref_u or nqp::bindpos_u variants 
at all.

At MoarVM level the atposref_i op is implemented by a call to 
MVM_nativeref_pos_i:
https://github.com/MoarVM/MoarVM/blob/
0a62f0c6a0b0696b6894e23fcac8431fdbd49db2/src/core/interp.c#L4656

This function then creates an object of type MVM_hll_current(tc)->int_pos_ref 
which in Raku is an IntPosRef object defined in:
https://github.com/rakudo/rakudo/blob/
168da39ae24b84f78525280a8d6750dd3cf1a176/src/Perl6/bootstrap.c/
BOOTSTRAP.nqp#L1747

As you see, the base type there is int, i.e. signed.

So to get a proper unsigned int out of CArray (and others), we need to extend 
the HLL config with uint_lex_ref, uint_attr_ref, uint_pos_ref and 
uint_multidim_ref fields, add corresponding UIntLexRef, UIntAttrRef, 
UIntPosRef and UIntMultiDimRef types to Rakudo's bootstrap and add unsigned 
versions of various ops like atposref_u, getattrref_u and so on.

The good news is that none of this is difficult. It's mostly copy & paste with 
adjustments. The bad news is that this is a bit like character encoding 
issues, i.e. you have to get the whole chain from input to final output right 
to get correct results. It may happen that your change is entirely correct but 
still shows no effect, because there's some other stage that's still handling 
it wrong.

Anyway, I'd love for someone to finally get our native unsigned support into 
shape!

**Comment by niner** [comment]
This is now fixed as of https://github.com/rakudo/rakudo/commit/8bba2aec43c14ad9a2ce5a00d2333b46cad30c14
Would you be so kind to contribute a PR with a test for this to rakudo?

**Comment by lizmat** [comment]
Test added with https://github.com/rakudo/rakudo/commit/ccf01a599b
